File: typescript/packages/bench/src/services/customerRetentionService.ts
/** Repeat-business share. */

import type { Dataset } from "../dataset";
import { AbstractService } from "../support/abstractService";

/** Repeat-business share. */
export class CustomerRetentionService extends AbstractService {
    /** Service name, used as the audit actor. */
    static readonly NAME = "customer-retention";

    constructor() {
        super(CustomerRetentionService.NAME);
    }

    /** Share of customers with more than one order. */
    rate(data: Dataset): number {
        if (data.customers.length === 0) {
            return 0;
        }

        const repeat = data.customers.filter((customer) => data.ordersFor(customer.id).length > 1).length;

        return repeat / data.customers.length;
    }
}

File: rust/crates/atelier-bench/src/services/customer_retention_service.rs
//! Repeat-business share.

use crate::dataset::Dataset;
use crate::support::abstract_service::ServiceBase;

/// Repeat-business share.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CustomerRetentionService {
    base: ServiceBase,
}

impl CustomerRetentionService {
    /// Service name, used as the audit actor.
    pub const NAME: &'static str = "customer-retention";

    pub fn new() -> Self {
        Self {
            base: ServiceBase::new(Self::NAME),
        }
    }

    /// Shared audit trail.
    pub fn base(&self) -> &ServiceBase {
        &self.base
    }

    /// Mutable access so callers can record their own actions.
    pub fn base_mut(&mut self) -> &mut ServiceBase {
        &mut self.base
    }

    /// Share of customers with more than one order.
    pub fn rate(&self, data: &Dataset) -> f64 {
        if data.customers.is_empty() {
            return 0.0;
        }

        let repeat = data
            .customers
            .iter()
            .filter(|customer| data.orders_for(customer.id).len() > 1)
            .count();

        repeat as f64 / data.customers.len() as f64
    }
}

impl Default for CustomerRetentionService {
    fn default() -> Self {
        Self::new()
    }
}
